test/common/upstream/health_check_fuzz.cc

```diff
@@ -112,34 +111,402 @@ void HealthCheckFuzz::triggerTimeoutTimer(bool last_action) {
+    return response;
+  }
+  case test::common::upstream::GrpcRespondBytes::kData: {
+    // Arbitrarily Generated Bytes
+    Buffer::OwnedImpl response;
+    response.add(grpc_respond_bytes.data());
+
+    ENVOY_LOG_MISC(trace,
+                   "Responded with {}. Length (in bytes) = {}. These bytes were arbitrarily "
+                   "generated and have no structure.",
+                   grpc_respond_bytes.data(), grpc_respond_bytes.data().length());
+
+    return response;
+  }
+  default: { // shouldn't hit
+    NOT_REACHED_GCOVR_EXCL_LINE;
+    break;
+  }
+  }
+}
+
+// Logic from respondResponseSpec() in unit tests
+void GrpcHealthCheckFuzz::respond(test::common::upstream::GrpcRespond grpc_respond) {
+  if (!test_sessions_[0]->timeout_timer_->enabled_) {
+    ENVOY_LOG_MISC(trace, "Timeout timer is disabled. Skipping response.");
+    return;
+  }
+  // These booleans help figure out when to end the stream
+  bool has_data = false;
+  bool has_trailers = false;
+
+  if (grpc_respond.has_grpc_respond_bytes()) {
+    switch (grpc_respond.grpc_respond_bytes().grpc_respond_bytes_selector_case()) {
+    case test::common::upstream::GrpcRespondBytes::kStatus: {
+      has_data = true;
+      break;
+    }
+    case test::common::upstream::GrpcRespondBytes::kData: {
+      if (grpc_respond.grpc_respond_bytes().data().length() != 0) {
+        has_data = true;
+      }
+      break;
+    }
+    default: // If the fuzzer generates an empty grpc_respond_bytes message
+      break;
+    }
+  }
+  // TODO: Should I hardcode trailers? Particularly grpc-status = 0?
+  if (grpc_respond.has_grpc_respond_trailers()) {
+    if (grpc_respond.grpc_respond_trailers().has_trailers()) {
+      if (grpc_respond.grpc_respond_trailers().trailers().headers_size() != 0) {
+        has_trailers = true;
+      }
+    }
+  }
+
+  ENVOY_LOG_MISC(trace, "Has data: {}. Has trailers: {}.", has_data, has_trailers);
+
+  bool end_stream_on_headers = !has_data && !has_trailers;
+
+  std::unique_ptr<Http::TestResponseHeaderMapImpl> response_headers =
+      std::make_unique<Http::TestResponseHeaderMapImpl>(
+          Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(
+              grpc_respond.grpc_respond_headers().headers(), {}, {}));
+
+  response_headers->setStatus("200"); // Required for gRPC
+
+  ENVOY_LOG_MISC(trace, "Responded headers {}", *response_headers.get());
+  test_sessions_[0]->stream_response_callbacks_->decodeHeaders(std::move(response_headers),
+                                                               end_stream_on_headers);
+
+  // If the interval timer is enabled, that means that the rpc is complete, as decodeHeaders hit a
+  // certain branch that called onRpcComplete(), logically representing a completed rpc call. Thus,
+  // skip the next responses until explicility invoking interval timer as cleanup.
+  if (has_data && !test_sessions_[0]->interval_timer_->enabled_) {
+    bool end_stream_on_data = !has_trailers;
```

Comment:
you can choose to keep this for clarity, but you could always pass !has_trailers to decodeData

Reply:
I thought about this too. Since all this logic is pulled from unit tests and they have another boolean only for clarity, I decided to keep it as such.